logger/logger.go

```diff
@@ -49,73 +20,40 @@ type Logger struct {
 
 // Message encapsulates app, level, message and options data.
 type Message struct {
-	App     string
 	Body    string
-	Level   string
 	Options Options
 }
 
-// InvalidOptionMessage represents an issue with the supplied configuration.
-type InvalidOptionMessage struct {
-	Option  string
-	Message string
+// Payload contains the properties sent to the ingestion endpoint
+type Payload struct {
+	APIKey     string `json:"apikey,omitempty"`
+	Hostname   string `json:"hostname,omitempty"`
+	IPAddress  string `json:"ip,omitempty"`
+	MacAddress string `json:"mac,omitempty"`
+	Tags       string `json:"tags,omitempty"`
+	Lines      []Line `json:"lines,omitempty"`
 }
 
-func (e InvalidOptionMessage) String() string {
-	return fmt.Sprintf("Options.%s: %s", e.Option, e.Message)
+// Line contains properties related to an individual log message
+type Line struct {
+	App   string       `json:"app,omitempty"`
+	Body  string       `json:"line,omitempty"`
+	Level string       `json:"level,omitempty"`
+	Env   string       `json:"env,omitempty"`
+	Meta  metaEnvelope `json:"meta,omitempty"`
 }
 
-func (options *Options) validate() error {
-	var problems []string
-	reMacAddress := regexp.MustCompile(`^([0-9a-fA-F][0-9a-fA-F]:){5}([0-9a-fA-F][0-9a-fA-F])`)
-	reHostname := regexp.MustCompile(`(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9-]*[A-Za-z0-9])`)
-	reIPAddress := regexp.MustCompile(`(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)(\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}`)
-
-	validateOptionLength("App", options.App, &problems)
-	validateOptionLength("Env", options.Env, &problems)
-	validateOptionLength("Hostname", options.Hostname, &problems)
-	validateOptionLength("Level", options.Level, &problems)
-
-	if options.MacAddress != "" && (!reMacAddress.MatchString(options.MacAddress)) {
-		problems = append(problems, InvalidOptionMessage{"MacAddress", "invalid format"}.String())
-	}
-
-	if options.Hostname != "" && !reHostname.MatchString(options.Hostname) {
-		problems = append(problems, InvalidOptionMessage{"Hostname", "invalid format"}.String())
-	}
-
-	if options.IPAddress != "" && !reIPAddress.MatchString(options.IPAddress) {
-		problems = append(problems, InvalidOptionMessage{"IPAddress", "invalid format"}.String())
-	}
-
-	if options.FlushInterval == 0 {
-		options.FlushInterval = 5 * time.Second
-	}
-	if options.MaxBufferLen == 0 {
-		options.MaxBufferLen = 5
-	}
-	if options.IngestURL == "" {
-		options.IngestURL = defaultIngestURL
-	}
-
-	if len(problems) > 0 {
-		return errors.New(strings.Join(problems, ", "))
-	}
-
-	return nil
+type metaEnvelope struct {
+	Indexed bool
+	Meta    string
 }
 
-func validateOptionLength(option string, value string, problems *[]string) {
-	if len(value) > 32 {
-		*problems = append(*problems, InvalidOptionMessage{option, "length must be less than 32"}.String())
+func (me metaEnvelope) MarshalJSON() ([]byte, error) {
+	if !me.Indexed {
```

Comment:
Does this Indexed boolean have to do with index_meta?

Reply:
yes - the `MetaEnvelope` is a wrapper struct that allows for the raw JSON to either remain in stringified form when `Indexed == false`, or get marshalled into the object structure and indexed on our end.

in the previous version of this code this handling was not quite working when `IndexMeta` was false, because after getting stringified it would get un-stringified [here ](https://github.com/logdna/logdna-go/pull/10/files#diff-ed0770fdbf87b0c6d536e33a99a8df9cL306)